core/src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -1800,86 +1875,42 @@ wordWithQuotesWith isInvalid keepQuotes isEsc toRight isSep
--- this parenthesis ) is ignored.
--- * Brackets can be nested.
+-- | 'wordWithQuotes' without processing the quotes and escape function
+-- supplied to escape the quote char within a quote. Can be used to parse words
+-- keeping the quotes and escapes intact.
 --
--- Examples of quotes - "double quotes", 'single quotes', (parens), {braces},
--- ((nested) brackets).
+-- >>> wordKeepQuotes = Parser.wordWithQuotes True (\_ _ -> Nothing)
 --
--- Example:
---
--- >>> :{
--- >>> q x =
--- >>>     case x of
--- >>>         '"' -> Just x
--- >>>         '\'' -> Just x
--- >>>         _ -> Nothing
--- >>> :}
---
--- >>> p = Parser.wordWithQuotes (== '\\') q isSpace Fold.toList
--- >>> Stream.parse p $ Stream.fromList "a\"b'c\";'d\"e'f ghi"
--- Right "a\"b'c\";'d\"e'f"
---
--- Note that outer quotes and backslashes from the input string are consumed by
--- Haskell, therefore, the actual input string passed to the parser is:
--- a"b'c";'d"e'f ghi
---
--- Similarly, when printing, double quotes are escaped by Haskell.
---
-{-# INLINE wordWithQuotes #-}
-wordWithQuotes :: (Monad m, Eq a) =>
-       (a -> Bool)     -- ^ Matches an escape elem?
+{-# INLINE wordKeepQuotes #-}
+wordKeepQuotes :: (Monad m, Eq a) =>
+       a               -- ^ Escape char
     -> (a -> Maybe a)  -- ^ If left quote, return right quote, else Nothing.
     -> (a -> Bool)     -- ^ Matches a word separator?
     -> Fold m a b
     -> Parser a m b
-wordWithQuotes = wordWithQuotesWith (const False) True
+wordKeepQuotes =
+    -- Escape the quote char itself
+    wordWithQuotes True (\q x -> if q == x then Just x else Nothing)
 
--- | Like 'wordWithQuotes' but strips the quotes and escape chars.
---
--- Example:
---
--- >>> :{
--- >>> q x =
--- >>>     case x of
--- >>>         '"' -> Just x
--- >>>         '\'' -> Just x
--- >>>         _ -> Nothing
--- >>> :}
---
--- >>> p = Parser.wordStripQuotes (== '\\') q isSpace Fold.toList
--- >>> Stream.parse p $ Stream.fromList "a\"b'c\";'d\"e'f ghi"
--- Right "ab'c;d\"ef"
---
--- Note that outer quotes and backslashes from the input string are consumed by
--- Haskell, therefore, the actual input string passed to the parser is:
--- a"b'c";'d"e'f ghi
---
--- From this the outermost quotes are stripped by the parser, double quotes inside
--- single quotes or vice-versa are kept.
+-- See the "Quoting Rules" section in the "bash" manual page for a primer on
+-- how quotes are used by shells.
+
+-- | 'wordWithQuotes' with quote processing applied and escape function
+-- supplied to escape the quote char within a quote. Can be ysed to parse words
+-- and processing the quoting and escaping at the same time.
 --
--- Similarly, when printing, double quotes are escaped by Haskell.
+-- >>> wordProcessQuotes = Parser.wordWithQuotes False (\_ _ -> Nothing)
 --
-{-# INLINE wordStripQuotes #-}
-wordStripQuotes :: (Monad m, Eq a) =>
-       (a -> Bool)     -- ^ Matches an escape elem?
+{-# INLINE wordProcessQuotes #-}
+wordProcessQuotes :: (Monad m, Eq a) =>
```

Comment:
I like `wordStripQuotes` over `wordProcessQuotes`.
By default, I can't think of process as utilizing the quotes.
I guess you can use `wordUtilizeQuotes`.
I guess process is also fine.